src/clj/game/macros.clj

```diff
@@ -26,6 +26,12 @@
       rd-runnable (not (:rd (get-in (:runner @state) [:register :cannot-run-on-server])))
       archives-runnable (not (:archives (get-in (:runner @state) [:register :cannot-run-on-server])))
       tagged (is-tagged? state)
+      ;; true if the current run (or the previous run if no current run) was initiated by this card
+      this-card-run (and (get-in card [:special :run-id])
+                         (= (get-in card [:special :run-id])
+                            (if (:run @state)
+                              (:run-id (:run @state))
+                              (:run-id (get-in (:runner @state) [:register :last-run])))))
```

Comment:
Is this a good idea? I realize it would be much more annoying to do it another way, but what about checking the `run` passed into `:run-ends`? I see that we're only passing the server into `:successful-run`, which sucks lol. jk This is a fine solution for now. We can fix `:successful-run` and move to something that isn't so stateful another time.

Reply:
I ended up reworking :successful-run and friends to take the run dict as target instead, making them consistent with :run-ends and :unsuccessful-run. It made the PR bigger than it used to, but it should alleviate some of the (justified!) concerns that this would cause issues.